In [1]:
!pip3 install openpyxl

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
uriage_data = pd.read_csv('E:/data/サンプルコード_20201021/2章/uriage.csv')
uriage_data.head()

purchase_date item_name  item_price customer_name
0  2019-06-13 18:02:34       商品A       100.0         深井菜々美
1  2019-07-13 13:05:29     商 品 S         NaN          浅田賢二
2  2019-05-11 19:42:07     商 品 a         NaN          南部慶二
3  2019-02-12 23:40:45       商品Z      2600.0          麻生莉緒
4  2019-04-22 03:09:35       商品a         NaN          平田鉄二

In [4]:
kokyaku_data = pd.read_excel("E:/data/サンプルコード_20201021/2章/kokyaku_daicho.xlsx")
kokyaku_data.head()

顧客名        かな  地域                     メールアドレス         登録日
0   須賀ひとみ    すが ひとみ  H市     suga_hitomi@example.com  2018/01/04
1  岡田　 敏也   おかだ としや  E市   okada_toshiya@example.com       42782
2    芳賀 希    はが のぞみ  A市     haga_nozomi@example.com  2018/01/07
3   荻野  愛    おぎの あい  F市        ogino_ai@example.com       42872
4   栗田 憲一  くりた けんいち  E市  kurita_kenichi@example.com       43127

In [5]:
uriage_data["purchase_date"] = pd.to_datetime(uriage_data["purchase_date"])
uriage_data["purchase_month"] = uriage_data["purchase_date"].dt.strftime("%Y%m")
res = uriage_data.pivot_table(index      = "purchase_month",
                              columns    = "item_name",
                              aggfunc    = "size",
                              fill_value = 0)
res

item_name         商品W   商 品 n   商品E   商品M   商品P   商品S   商品W   商品X  商  品O  \
purchase_month                                                             
201901              0       1     0     0     0     0     0     0      0   
201902              0       0     0     0     0     0     0     1      0   
201903              0       0     1     1     1     0     0     0      0   
201904              1       0     0     0     0     0     0     0      0   
201905              0       0     0     0     0     1     0     0      0   
201906              0       0     0     0     0     0     1     0      0   
201907              0       0     0     0     0     0     0     0      1   

item_name       商  品Q  ...  商品k  商品l  商品o  商品p  商品r  商品s  商品t  商品v  商品x  商品y  
purchase_month         ...                                                    
201901              0  ...    1    1    1    0    0    0    0    0    0    0  
201902              0  ...    0    0    0    0    0    1    1    1    0    0  
201903              0  ...    0    0    0    0    0    0    0    0    0    0  
201904              1  ...    0    0    0    0    0    1    0    0    0    0  
201905              0  ...    0    1    0    0    0    0    0    0    0    1  
201906              0  ...    0    0    0    1    0    0    0    0    1    0  
201907              0  ...    0    0    1    0    2    0    0    0    0    0  

[7 rows x 99 columns]

In [6]:
res = uriage_data.pivot_table(index      = "purchase_month",
                              columns    = "item_name",
                              values     = "item_price",
                              aggfunc    = "sum",
                              fill_value = 0)
res

item_name         商品W   商 品 n   商品E   商品M   商品P   商品S   商品W   商品X  商  品O  \
purchase_month                                                             
201901              0    1400     0     0     0     0     0     0      0   
201902              0       0     0     0     0     0     0  2400      0   
201903              0       0   500  1300  1600     0     0     0      0   
201904           2300       0     0     0     0     0     0     0      0   
201905              0       0     0     0     0  1900     0     0      0   
201906              0       0     0     0     0     0  2300     0      0   
201907              0       0     0     0     0     0     0     0      0   

item_name       商  品Q  ...   商品k   商品l   商品o   商品p   商品r   商品s   商品t   商品v  \
purchase_month         ...                                                   
201901              0  ...  1100  1200  1500     0     0     0     0     0   
201902              0  ...     0     0     0     0     0  1900  2000  2200   
201903              0  ...     0     0     0     0     0     0     0     0   
201904           1700  ...     0     0     0     0     0  1900     0     0   
201905              0  ...     0  1200     0     0     0     0     0     0   
201906              0  ...     0     0     0  1600     0     0     0     0   
201907              0  ...     0     0  1500     0  1800     0     0     0   

item_name        商品x   商品y  
purchase_month              
201901             0     0  
201902             0     0  
201903             0     0  
201904             0     0  
201905             0  2500  
201906          2400     0  
201907             0     0  

[7 rows x 99 columns]

In [7]:
print(len(pd.unique(uriage_data.item_name)))
uriage_data["item_name"] = uriage_data["item_name"].str.upper()
uriage_data["item_name"] = uriage_data["item_name"].str.replace("  ", "")
uriage_data["item_name"] = uriage_data["item_name"].str.replace(" ", "")
uriage_data.sort_values(by = ["item_name"],
                        ascending = True)

99


purchase_date item_name  item_price customer_name purchase_month
0    2019-06-13 18:02:34       商品A       100.0         深井菜々美         201906
1748 2019-05-19 20:22:22       商品A       100.0          松川綾女         201905
223  2019-06-25 08:13:20       商品A       100.0           板橋隆         201906
1742 2019-06-13 16:03:17       商品A       100.0          小平陽子         201906
1738 2019-02-10 00:28:43       商品A       100.0          松田浩正         201902
...                  ...       ...         ...           ...            ...
2880 2019-04-22 00:36:52       商品Y         NaN          田辺光洋         201904
2881 2019-04-30 14:21:09       商品Y         NaN          高原充則         201904
1525 2019-01-24 10:27:23       商品Y      2500.0         五十嵐春樹         201901
1361 2019-05-28 13:45:32       商品Y      2500.0         大崎ヒカル         201905
3    2019-02-12 23:40:45       商品Z      2600.0          麻生莉緒         201902

[2999 rows x 5 columns]

In [8]:
print(pd.unique(uriage_data["item_name"]))
print(len(pd.unique(uriage_data.item_name)))

['商品A' '商品S' '商品Z' '商品V' '商品O' '商品U' '商品L' '商品C' '商品I' '商品R' '商品X' '商品G'
 '商品P' '商品Q' '商品Y' '商品N' '商品W' '商品E' '商品K' '商品B' '商品F' '商品D' '商品M' '商品H'
 '商品T' '商品J']
26


In [9]:
uriage_data.isnull().sum()
# uriage_data.isnull().any(axis=0)

purchase_date       0
item_name           0
item_price        387
customer_name       0
purchase_month      0
dtype: int64

In [10]:
flg_is_null = uriage_data["item_price"].isnull()
tes = uriage_data.loc[flg_is_null,"item_name"].unique()
tes
# for trg in list(uriage_data.loc[flg_is_null,
#                                 "item_name"].unique()):
#     price = uriage_data.loc[(~flg_is_null) & (uriage_data["item_name"] == trg),
#                             "item_price"].max()
#     uriage_data["item_price"].loc[(flg_is_null) & (uriage_data["item_name"] == trg)] = price
# uriage_data.head()

array(['商品S', '商品A', '商品P', '商品N', '商品W', '商品R', '商品I', '商品L', '商品F',
       '商品O', '商品B', '商品C', '商品V', '商品Q', '商品U', '商品K', '商品T', '商品X',
       '商品E', '商品M', '商品G', '商品J', '商品D', '商品H', '商品Y'], dtype=object)

金額欠損の確認

In [11]:
uriage_data.isnull().any(axis=0)

purchase_date     False
item_name         False
item_price         True
customer_name     False
purchase_month    False
dtype: bool

金額欠損の補完
商品名が同じものは全て同じ値段に変更（データ期間内に値段変更がない前提のため）

In [12]:
flg_is_null = uriage_data["item_price"].isnull()                                                   #item_priceが欠損してる行を格納
for trg in list(uriage_data.loc[flg_is_null, "item_name"].unique()):                               #欠損してる行のitem_nameを取得してループ
    price = uriage_data.loc[(~flg_is_null) & (uriage_data["item_name"] == trg),"item_price"].max() #欠損してない行でitem_naemが一致したものがある時、その金額を一意にして取得 （このままだとコピー値の変更による警告なので以下で回避）
    for index,flg in enumerate(flg_is_null):
        if flg and uriage_data.loc[index,"item_name"] == trg:
            uriage_data.loc[index,"item_price"] = price
#     uriage_data["item_price"].loc[(flg_is_null) & (uriage_data["item_name"]==trg)] = price         #取得した金額を欠損のある行で、item_nameが一致してるところに入れる
uriage_data


purchase_date item_name  item_price customer_name purchase_month
0    2019-06-13 18:02:34       商品A       100.0         深井菜々美         201906
1    2019-07-13 13:05:29       商品S      1900.0          浅田賢二         201907
2    2019-05-11 19:42:07       商品A       100.0          南部慶二         201905
3    2019-02-12 23:40:45       商品Z      2600.0          麻生莉緒         201902
4    2019-04-22 03:09:35       商品A       100.0          平田鉄二         201904
...                  ...       ...         ...           ...            ...
2994 2019-02-15 02:56:39       商品Y      2500.0          福島友也         201902
2995 2019-06-22 04:03:43       商品M      1300.0          大倉晃司         201906
2996 2019-03-29 11:14:05       商品Q      1700.0          尾形小雁         201903
2997 2019-07-14 12:56:49       商品H       800.0          芦田博之         201907
2998 2019-07-21 00:31:36       商品D       400.0          石田郁恵         201907

[2999 rows x 5 columns]

検証 nullが消えたか

In [13]:
uriage_data.isnull().any(axis=0)

purchase_date     False
item_name         False
item_price        False
customer_name     False
purchase_month    False
dtype: bool

検証 金額が正しく補完されたか

In [14]:
for trg in list(uriage_data["item_name"].sort_values().unique()):   #全ての商品名でループ
    #商品名が一致したデータの最大値と最小値の表示＋nullチェック
    print(trg + "の最大額：" + str(uriage_data.loc[uriage_data["item_name"]==trg]["item_price"].max()) +
          "の最少額：" + str(uriage_data.loc[uriage_data["item_name"]==trg]["item_price"].min(skipna = False)))

商品Aの最大額：100.0の最少額：100.0
商品Bの最大額：200.0の最少額：200.0
商品Cの最大額：300.0の最少額：300.0
商品Dの最大額：400.0の最少額：400.0
商品Eの最大額：500.0の最少額：500.0
商品Fの最大額：600.0の最少額：600.0
商品Gの最大額：700.0の最少額：700.0
商品Hの最大額：800.0の最少額：800.0
商品Iの最大額：900.0の最少額：900.0
商品Jの最大額：1000.0の最少額：1000.0
商品Kの最大額：1100.0の最少額：1100.0
商品Lの最大額：1200.0の最少額：1200.0
商品Mの最大額：1300.0の最少額：1300.0
商品Nの最大額：1400.0の最少額：1400.0
商品Oの最大額：1500.0の最少額：1500.0
商品Pの最大額：1600.0の最少額：1600.0
商品Qの最大額：1700.0の最少額：1700.0
商品Rの最大額：1800.0の最少額：1800.0
商品Sの最大額：1900.0の最少額：1900.0
商品Tの最大額：2000.0の最少額：2000.0
商品Uの最大額：2100.0の最少額：2100.0
商品Vの最大額：2200.0の最少額：2200.0
商品Wの最大額：2300.0の最少額：2300.0
商品Xの最大額：2400.0の最少額：2400.0
商品Yの最大額：2500.0の最少額：2500.0
商品Zの最大額：2600.0の最少額：2600.0


15.顧客名の揺れを補正 揺れの状況確認
結合時のキーが顧客名しかなないので必須

In [15]:
kokyaku_data["顧客名"].head()

0     須賀ひとみ
1    岡田　 敏也
2      芳賀 希
3     荻野  愛
4     栗田 憲一
Name: 顧客名, dtype: object

In [16]:
uriage_data["customer_name"].head()

0    深井菜々美
1     浅田賢二
2     南部慶二
3     麻生莉緒
4     平田鉄二
Name: customer_name, dtype: object

姓名の間にスペース（半角、全角）がある

In [17]:
kokyaku_data["顧客名"] = kokyaku_data["顧客名"].str.replace("　", "") #全角スペースを削除
kokyaku_data["顧客名"] = kokyaku_data["顧客名"].str.replace(" ", "")  #半角スペースを削除
kokyaku_data["顧客名"].head()

0    須賀ひとみ
1     岡田敏也
2      芳賀希
3      荻野愛
4     栗田憲一
Name: 顧客名, dtype: object

16.日付の揺れの補正 揺れを確認する

日付ではない数字がある。<br>
2018/01/04<br>
42782<br>
原因は取り込み元のデータ（Excel）<br>
2018/01/04  （フォント違い）<br>
2017年2月16日<br>
2018/01/07<br><br>
数字は1900/01/01をスタートにした経過日数（シリアル値）で、エクセルと比べて-2日ずれて変換されている<br>
１．エクセルの最小日付は1900/01/01でシリアル値が1だが、pythonでは0と認識する<br>
２．エクセルではうるう年でない日付もシリアル値で有効になるが、pythonでは無効となるためカウントされない

In [18]:
flg_is_serial = kokyaku_data["登録日"].astype("str").str.isdigit() #strにキャストして、数値の行を取り出す
flg_is_serial.sum()

22

In [19]:
#数値の行は経過時間として1900/01/01をプラス
fromSerial = pd.to_timedelta(kokyaku_data.loc[flg_is_serial, "登録日"].astype("float")-2,unit="D") + pd.to_datetime("1900/01/01") #-2はエクセルのシリアル値の対応
fromSerial

1     2017-02-16
3     2017-05-17
4     2018-01-27
21    2017-07-04
27    2017-06-15
47    2017-01-06
49    2017-07-13
53    2017-04-08
76    2018-03-29
80    2018-01-10
99    2017-05-30
114   2018-06-03
118   2018-01-29
122   2018-04-16
139   2017-05-25
143   2017-03-24
155   2017-01-19
172   2018-03-22
179   2017-01-08
183   2017-07-24
186   2018-07-13
192   2018-06-08
Name: 登録日, dtype: datetime64[ns]

In [20]:
#その他の日付も書式統一
fromString = pd.to_datetime(kokyaku_data.loc[~flg_is_serial, "登録日"])
fromString

0     2018-01-04
2     2018-01-07
5     2017-06-20
6     2018-06-11
7     2017-05-19
         ...    
195   2017-06-20
196   2018-06-20
197   2017-04-29
198   2019-04-19
199   2019-04-23
Name: 登録日, Length: 178, dtype: datetime64[ns]

In [21]:
kokyaku_data["登録日"] = pd.concat([fromSerial, fromString]) #日付を結合して変更
kokyaku_data

顧客名           かな  地域                         メールアドレス        登録日
0    須賀ひとみ       すが ひとみ  H市         suga_hitomi@example.com 2018-01-04
1     岡田敏也      おかだ としや  E市       okada_toshiya@example.com 2017-02-16
2      芳賀希       はが のぞみ  A市         haga_nozomi@example.com 2018-01-07
3      荻野愛       おぎの あい  F市            ogino_ai@example.com 2017-05-17
4     栗田憲一     くりた けんいち  E市      kurita_kenichi@example.com 2018-01-27
..     ...          ...  ..                             ...        ...
195   川上りえ      かわかみ りえ  G市        kawakami_rie@example.com 2017-06-20
196   小松季衣      こまつ としえ  E市      komatsu_toshie@example.com 2018-06-20
197   白鳥りえ      しらとり りえ  F市       shiratori_rie@example.com 2017-04-29
198  大西隆之介  おおにし りゅうのすけ  H市  oonishi_ryuunosuke@example.com 2019-04-19
199   福井美希       ふくい みき  D市         fukui_miki1@example.com 2019-04-23

[200 rows x 5 columns]

揺れが消えたか検証１ データ件数でチェック

In [22]:
kokyaku_data["登録年月"] = kokyaku_data["登録日"].dt.strftime("%Y%m") # 登録日を年月に変換
rslt = kokyaku_data.count()["顧客名"] # 合計数を出す
print(rslt)
rslt = kokyaku_data.groupby("登録年月").count()["顧客名"] # groupbyで各顧客名の件数を取得
print(rslt)
print(len(kokyaku_data)) #データ全体の件数と合ってるか

200
登録年月
201701    15
201702    11
201703    14
201704    15
201705    14
201706    13
201707    17
201801    13
201802    15
201803    17
201804     5
201805    19
201806    13
201807    17
201904     2
Name: 顧客名, dtype: int64
200


検証２ 数値データが残ってないかチェック

In [23]:
flg_is_serial = kokyaku_data["登録日"].astype("str").str.isdigit()
flg_is_serial.sum()

0

１８、customer_nameと顧客名をキーに売上履歴と顧客台帳でJOIN（IDなどの共通キーがない）

In [24]:
uriage_data

purchase_date item_name  item_price customer_name purchase_month
0    2019-06-13 18:02:34       商品A       100.0         深井菜々美         201906
1    2019-07-13 13:05:29       商品S      1900.0          浅田賢二         201907
2    2019-05-11 19:42:07       商品A       100.0          南部慶二         201905
3    2019-02-12 23:40:45       商品Z      2600.0          麻生莉緒         201902
4    2019-04-22 03:09:35       商品A       100.0          平田鉄二         201904
...                  ...       ...         ...           ...            ...
2994 2019-02-15 02:56:39       商品Y      2500.0          福島友也         201902
2995 2019-06-22 04:03:43       商品M      1300.0          大倉晃司         201906
2996 2019-03-29 11:14:05       商品Q      1700.0          尾形小雁         201903
2997 2019-07-14 12:56:49       商品H       800.0          芦田博之         201907
2998 2019-07-21 00:31:36       商品D       400.0          石田郁恵         201907

[2999 rows x 5 columns]

In [25]:
kokyaku_data

顧客名           かな  地域                         メールアドレス        登録日    登録年月
0    須賀ひとみ       すが ひとみ  H市         suga_hitomi@example.com 2018-01-04  201801
1     岡田敏也      おかだ としや  E市       okada_toshiya@example.com 2017-02-16  201702
2      芳賀希       はが のぞみ  A市         haga_nozomi@example.com 2018-01-07  201801
3      荻野愛       おぎの あい  F市            ogino_ai@example.com 2017-05-17  201705
4     栗田憲一     くりた けんいち  E市      kurita_kenichi@example.com 2018-01-27  201801
..     ...          ...  ..                             ...        ...     ...
195   川上りえ      かわかみ りえ  G市        kawakami_rie@example.com 2017-06-20  201706
196   小松季衣      こまつ としえ  E市      komatsu_toshie@example.com 2018-06-20  201806
197   白鳥りえ      しらとり りえ  F市       shiratori_rie@example.com 2017-04-29  201704
198  大西隆之介  おおにし りゅうのすけ  H市  oonishi_ryuunosuke@example.com 2019-04-19  201904
199   福井美希       ふくい みき  D市         fukui_miki1@example.com 2019-04-23  201904

[200 rows x 6 columns]

In [26]:
join_data = pd.merge(uriage_data, kokyaku_data, left_on = "customer_name", right_on = "顧客名", how="left") #左外部結合
join_data = join_data.drop("customer_name",axis=1) #customer_name列削除 必要ないらしい
join_data

purchase_date item_name  item_price purchase_month    顧客名  \
0    2019-06-13 18:02:34       商品A       100.0         201906  深井菜々美   
1    2019-07-13 13:05:29       商品S      1900.0         201907   浅田賢二   
2    2019-05-11 19:42:07       商品A       100.0         201905   南部慶二   
3    2019-02-12 23:40:45       商品Z      2600.0         201902   麻生莉緒   
4    2019-04-22 03:09:35       商品A       100.0         201904   平田鉄二   
...                  ...       ...         ...            ...    ...   
2994 2019-02-15 02:56:39       商品Y      2500.0         201902   福島友也   
2995 2019-06-22 04:03:43       商品M      1300.0         201906   大倉晃司   
2996 2019-03-29 11:14:05       商品Q      1700.0         201903   尾形小雁   
2997 2019-07-14 12:56:49       商品H       800.0         201907   芦田博之   
2998 2019-07-21 00:31:36       商品D       400.0         201907   石田郁恵   

            かな  地域                       メールアドレス        登録日    登録年月  
0      ふかい ななみ  C市      fukai_nanami@example.com 2017-01-26  201701  
1      あさだ けんじ  C市       asada_kenji@example.com 2018-04-07  201804  
2      なんぶ けいじ  A市      nannbu_keiji@example.com 2018-06-19  201806  
3       あそう りお  D市          asou_rio@example.com 2018-07-22  201807  
4      ひらた てつじ  D市    hirata_tetsuji@example.com 2017-06-07  201706  
...        ...  ..                           ...        ...     ...  
2994  ふくしま ともや  B市  fukushima_tomoya@example.com 2017-07-01  201707  
2995  おおくら こうじ  E市      ookura_kouji@example.com 2018-03-31  201803  
2996   おがた こがん  B市       ogata_kogan@example.com 2017-03-15  201703  
2997  あしだ ひろゆき  E市   ashida_hiroyuki@example.com 2018-07-13  201807  
2998   いしだ いくえ  B市       ishida_ikue@example.com 2017-02-05  201702  

[2999 rows x 10 columns]

クレンジング終了<br>
CSVに列の配置を調整（関連の強いものを近くに移動）してダンプ

In [27]:
dump_data = join_data[["purchase_date", "purchase_month", "item_name", "item_price", "顧客名", "かな", "地域", "メールアドレス", "登録日"]]
dump_data

purchase_date purchase_month item_name  item_price    顧客名  \
0    2019-06-13 18:02:34         201906       商品A       100.0  深井菜々美   
1    2019-07-13 13:05:29         201907       商品S      1900.0   浅田賢二   
2    2019-05-11 19:42:07         201905       商品A       100.0   南部慶二   
3    2019-02-12 23:40:45         201902       商品Z      2600.0   麻生莉緒   
4    2019-04-22 03:09:35         201904       商品A       100.0   平田鉄二   
...                  ...            ...       ...         ...    ...   
2994 2019-02-15 02:56:39         201902       商品Y      2500.0   福島友也   
2995 2019-06-22 04:03:43         201906       商品M      1300.0   大倉晃司   
2996 2019-03-29 11:14:05         201903       商品Q      1700.0   尾形小雁   
2997 2019-07-14 12:56:49         201907       商品H       800.0   芦田博之   
2998 2019-07-21 00:31:36         201907       商品D       400.0   石田郁恵   

            かな  地域                       メールアドレス        登録日  
0      ふかい ななみ  C市      fukai_nanami@example.com 2017-01-26  
1      あさだ けんじ  C市       asada_kenji@example.com 2018-04-07  
2      なんぶ けいじ  A市      nannbu_keiji@example.com 2018-06-19  
3       あそう りお  D市          asou_rio@example.com 2018-07-22  
4      ひらた てつじ  D市    hirata_tetsuji@example.com 2017-06-07  
...        ...  ..                           ...        ...  
2994  ふくしま ともや  B市  fukushima_tomoya@example.com 2017-07-01  
2995  おおくら こうじ  E市      ookura_kouji@example.com 2018-03-31  
2996   おがた こがん  B市       ogata_kogan@example.com 2017-03-15  
2997  あしだ ひろゆき  E市   ashida_hiroyuki@example.com 2018-07-13  
2998   いしだ いくえ  B市       ishida_ikue@example.com 2017-02-05  

[2999 rows x 9 columns]

In [28]:
dump_data.to_csv("dump_data.csv", index=False) #CSVでダンプ

ダンプデータを取り込み

In [29]:
import_data = pd.read_csv("dump_data.csv")
import_data

purchase_date  purchase_month item_name  item_price    顧客名  \
0     2019-06-13 18:02:34          201906       商品A       100.0  深井菜々美   
1     2019-07-13 13:05:29          201907       商品S      1900.0   浅田賢二   
2     2019-05-11 19:42:07          201905       商品A       100.0   南部慶二   
3     2019-02-12 23:40:45          201902       商品Z      2600.0   麻生莉緒   
4     2019-04-22 03:09:35          201904       商品A       100.0   平田鉄二   
...                   ...             ...       ...         ...    ...   
2994  2019-02-15 02:56:39          201902       商品Y      2500.0   福島友也   
2995  2019-06-22 04:03:43          201906       商品M      1300.0   大倉晃司   
2996  2019-03-29 11:14:05          201903       商品Q      1700.0   尾形小雁   
2997  2019-07-14 12:56:49          201907       商品H       800.0   芦田博之   
2998  2019-07-21 00:31:36          201907       商品D       400.0   石田郁恵   

            かな  地域                       メールアドレス                  登録日  
0      ふかい ななみ  C市      fukai_nanami@example.com  2017-01-26 00:00:00  
1      あさだ けんじ  C市       asada_kenji@example.com  2018-04-07 00:00:00  
2      なんぶ けいじ  A市      nannbu_keiji@example.com  2018-06-19 00:00:00  
3       あそう りお  D市          asou_rio@example.com  2018-07-22 00:00:00  
4      ひらた てつじ  D市    hirata_tetsuji@example.com  2017-06-07 00:00:00  
...        ...  ..                           ...                  ...  
2994  ふくしま ともや  B市  fukushima_tomoya@example.com  2017-07-01 00:00:00  
2995  おおくら こうじ  E市      ookura_kouji@example.com  2018-03-31 00:00:00  
2996   おがた こがん  B市       ogata_kogan@example.com  2017-03-15 00:00:00  
2997  あしだ ひろゆき  E市   ashida_hiroyuki@example.com  2018-07-13 00:00:00  
2998   いしだ いくえ  B市       ishida_ikue@example.com  2017-02-05 00:00:00  

[2999 rows x 9 columns]

purchase_monthを縦軸に商品毎に集計<br>
以下売上金額、顧客、地域毎に集計

In [30]:
byItem = import_data.pivot_table(index="purchase_month", columns="item_name", aggfunc="size", fill_value=0)
byItem

item_name       商品A  商品B  商品C  商品D  商品E  商品F  商品G  商品H  商品I  商品J  ...  商品Q  \
purchase_month                                                    ...        
201901           18   13   19   17   18   15   11   16   18   17  ...   17   
201902           19   14   26   21   16   14   14   17   12   14  ...   22   
201903           17   21   20   17    9   27   14   18   12   16  ...   23   
201904           17   19   24   20   18   17   14   11   18   13  ...   20   
201905           24   14   16   14   19   18   23   15   16   11  ...   13   
201906           24   12   11   19   13   18   15   13   19   22  ...   15   
201907           20   20   17   17   12   17   19   19   19   23  ...   15   

item_name       商品R  商品S  商品T  商品U  商品V  商品W  商品X  商品Y  商品Z  
purchase_month                                               
201901           21   20   17    7   22   13   14   10    0  
201902           22   22   23   19   22   24   16   11    1  
201903           16   20   12   23   18   16   21   16    0  
201904           20   16   16   11   15   14   16   20    0  
201905           22   18   16   16    9   21   16   20    0  
201906           16   21   12   18   20   17   15   13    0  
201907           19   23   21   13   28   16   18   12    0  

[7 rows x 26 columns]

In [31]:
byPrice = import_data.pivot_table(index="purchase_month", columns="item_name", values="item_price", aggfunc="sum", fill_value=0)
byPrice

item_name        商品A   商品B   商品C   商品D   商品E    商品F    商品G    商品H    商品I  \
purchase_month                                                             
201901          1800  2600  5700  6800  9000   9000   7700  12800  16200   
201902          1900  2800  7800  8400  8000   8400   9800  13600  10800   
201903          1700  4200  6000  6800  4500  16200   9800  14400  10800   
201904          1700  3800  7200  8000  9000  10200   9800   8800  16200   
201905          2400  2800  4800  5600  9500  10800  16100  12000  14400   
201906          2400  2400  3300  7600  6500  10800  10500  10400  17100   
201907          2000  4000  5100  6800  6000  10200  13300  15200  17100   

item_name         商品J  ...    商品Q    商品R    商品S    商品T    商品U    商品V    商品W  \
purchase_month         ...                                                    
201901          17000  ...  28900  37800  38000  34000  14700  48400  29900   
201902          14000  ...  37400  39600  41800  46000  39900  48400  55200   
201903          16000  ...  39100  28800  38000  24000  48300  39600  36800   
201904          13000  ...  34000  36000  30400  32000  23100  33000  32200   
201905          11000  ...  22100  39600  34200  32000  33600  19800  48300   
201906          22000  ...  25500  28800  39900  24000  37800  44000  39100   
201907          23000  ...  25500  34200  43700  42000  27300  61600  36800   

item_name         商品X    商品Y   商品Z  
purchase_month                      
201901          33600  25000     0  
201902          38400  27500  2600  
201903          50400  40000     0  
201904          38400  50000     0  
201905          38400  50000     0  
201906          36000  32500     0  
201907          43200  30000     0  

[7 rows x 26 columns]

In [32]:
byCustomer = import_data.pivot_table(index="purchase_month", columns="顧客名",  aggfunc="size", fill_value=0)
byCustomer

顧客名             さだ千佳子  中仁晶  中田美智子  丸山光臣  久保田倫子  亀井一徳  五十嵐春樹  井上桃子  井口寛治  \
purchase_month                                                            
201901              3    1      4     2      2     0      5     3     3   
201902              9    1      2     2      1     4      2     1     0   
201903              1    2      1     6      1     4      3     3     2   
201904              0    3      1     2      0     2      2     0     3   
201905              3    2      5     2      4     1      2     1     3   
201906              1    3      0     4      1     1      1     2     2   
201907              3    0      3     2      5     3      5     2     5   

顧客名             井川真悠子  ...  香椎優一  高原充則  高梨結衣  高沢美咲  高田さんま  鳥居広司  鶴岡薫  麻生莉緒  \
purchase_month         ...                                                   
201901              1  ...     0     1     1     1      5     2    0     2   
201902              4  ...     4     0     3     2      0     1    2     4   
201903              2  ...     3     1     6     2      4     2    4     2   
201904              2  ...     2     4     2     3      4     3    2     1   
201905              3  ...     1     1     1     0      2     2    3     4   
201906              3  ...     7     3     0     2      1     0    2     1   
201907              5  ...     2     4     4     2      0     2    4     3   

顧客名             黄川田博之  黒谷長利  
purchase_month               
201901              2     5  
201902              0     1  
201903              2     1  
201904              2     0  
201905              4     1  
201906              2     4  
201907              4     1  

[7 rows x 199 columns]

In [33]:
byRegion = import_data.pivot_table(index="purchase_month", columns="地域", aggfunc="size", fill_value=0)
byRegion

地域              A市  B市  C市  D市  E市  F市  G市  H市
purchase_month                                
201901          59  55  72  34  49  57  49  42
201902          71  46  65  48  61  52  43  63
201903          64  52  57  43  52  59  51  59
201904          64  48  54  45  48  58  40  52
201905          57  52  68  48  59  65  35  43
201906          53  47  61  30  51  51  58  58
201907          76  53  61  42  54  64  47  54

集計期間で購入してないユーザがいないかチェックして完了

In [34]:
away_data = pd.merge(uriage_data, kokyaku_data, left_on="customer_name", right_on="顧客名", how="right")
away_data[away_data["purchase_date"].isnull()][["顧客名", "メールアドレス", "登録日"]]

顧客名                  メールアドレス        登録日
2999  福井美希  fukui_miki1@example.com 2019-04-23